File: src/main/java/com/github/tianchenghang/agent/ToolCallAgent.java
package com.github.tianchenghang.agent;

import cn.hutool.core.collection.CollUtil;
import cn.hutool.core.util.StrUtil;
import com.alibaba.cloud.ai.dashscope.chat.DashScopeChatOptions;
import com.github.tianchenghang.agent.model.AgentState;
import java.util.stream.Collectors;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.extern.slf4j.Slf4j;
import org.springframework.ai.chat.messages.AssistantMessage;
import org.springframework.ai.chat.messages.ToolResponseMessage;
import org.springframework.ai.chat.messages.UserMessage;
import org.springframework.ai.chat.model.ChatResponse;
import org.springframework.ai.chat.prompt.ChatOptions;
import org.springframework.ai.chat.prompt.Prompt;
import org.springframework.ai.model.tool.ToolCallingManager;
import org.springframework.ai.tool.ToolCallback;

// Base agent class for handling tool calls, implementing the think and act methods, and can be used
// as a parent class for creating instances.
@EqualsAndHashCode(callSuper = true)
@Data
@Slf4j
public class ToolCallAgent extends ReActAgent {

  // Available tools
  private final ToolCallback[] availableTools;

  // Stores the response results of tool calls
  private ChatResponse toolCallChatResponse;

  // Tool calling manager
  private final ToolCallingManager toolCallingManager;

  // Disable Spring AI's built-in tool calling mechanism and maintain options and message context
  // manually
  private final ChatOptions chatOptions;

  public ToolCallAgent(ToolCallback[] availableTools) {
    super();
    this.availableTools = availableTools;
    this.toolCallingManager = ToolCallingManager.builder().build();

    // Disable Spring AI's built-in tool calling mechanism and maintain options and message context
    // manually
    this.chatOptions =
        DashScopeChatOptions.builder().withInternalToolExecutionEnabled(false).build();
  }

  // Processes the current state and decides the next action
  @Override
  public boolean think() {
    // Validate the prompt and concatenate the user prompt
    if (StrUtil.isNotBlank(getNextStepPrompt())) {
      var userMessage = new UserMessage(getNextStepPrompt());
      getMessageList().add(userMessage);
    }
    // Call the AI model to get the tool call results
    var messageList = getMessageList();
    var prompt = new Prompt(messageList, this.chatOptions);
    try {
      var chatResponse =
          getChatClient()
              .prompt(prompt)
              .system(getSystemPrompt())
              .tools(availableTools)
              .call()
              .chatResponse();
      // Record the response for later use in Act
      this.toolCallChatResponse = chatResponse;
      var assistantMessage = chatResponse.getResult().getOutput();
      var toolCallList = assistantMessage.getToolCalls();
      var result = assistantMessage.getText();
      log.info(getName() + " thinking: " + result);
      log.info(getName() + " selected " + toolCallList.size() + " tools to use");
      // Send the thinking content to SSE
      emitThinking(result);

      // If no tools need to be called, return false
      if (toolCallList.isEmpty()) {
        // Only when no tools are called, manually record the assistant message
        getMessageList().add(assistantMessage);
        return false;
      }

      var toolCallInfo =
          toolCallList.stream()
              .map(
                  toolCall ->
                      String.format(
                          "Tool name: %s, arguments: %s", toolCall.name(), toolCall.arguments()))
              .collect(Collectors.joining("\n"));
      log.info("Tool call information: {}", toolCallInfo);
      // Send tool call information to SSE
      for (var toolCall : toolCallList) {
        emitToolCall(toolCall.name(), toolCall.arguments());
      }
      // When tools need to be called, no need to record the assistant message as it will be
      // automatically recorded during tool calls
      return true;
    } catch (Exception e) {
      log.error(getName() + " process error: " + e.getMessage());
      getMessageList().add(new AssistantMessage(getName() + " process error: " + e.getMessage()));
      return false;
    }
  }

  @Override
  public String act() {
    if (!toolCallChatResponse.hasToolCalls()) {
      return "No tools need to be called";
    }
    // Call tools
    var prompt = new Prompt(getMessageList(), this.chatOptions);
    var toolExecutionResult = toolCallingManager.executeToolCalls(prompt, toolCallChatResponse);
    setMessageList(toolExecutionResult.conversationHistory());
    var toolResponseMessage =
        (ToolResponseMessage) CollUtil.getLast(toolExecutionResult.conversationHistory());
    // Check if a terminate tool was called
    var terminateToolCalled =
        toolResponseMessage.getResponses().stream()
            .anyMatch(response -> response.name().equals("doTerminate"));
    if (terminateToolCalled) {
      // Task completed, update state
      setState(AgentState.FINISHED);
    }
    // Send each tool's result to SSE
    for (ToolResponseMessage.ToolResponse response : toolResponseMessage.getResponses()) {
      emitToolResult(response.name(), response.responseData());
    }
    var results =
        toolResponseMessage.getResponses().stream()
            .map(
                response ->
                    "Tool " + response.name() + " returned result: " + response.responseData())
            .collect(Collectors.joining("\n"));
    log.info(results);
    return results;
  }
}

File: server/src/agent/tool-call-agent.ts
import { ToolCall } from '@langchain/core/messages/tool';
import { AIMessage, HumanMessage, ToolMessage } from '@langchain/core/messages';
import { ChatResponse } from '../llm/chat-model';
import { ToolExecutor } from '../tools/types';
import { TerminateTool } from '../tools/terminate-tool';
import { AgentState } from './model/agent-state.enum';
import { ReActAgent } from './re-act-agent';
import { StructuredTool } from '@langchain/core/tools';

export class ToolCallAgent extends ReActAgent {
  // Available tools
  protected availableTools: StructuredTool[] = [];

  // Stores the response result of tool calls
  protected toolCallResponse: ChatResponse | null = null;

  // Tool executor
  protected toolExecutor: ToolExecutor;

  constructor(tools: StructuredTool[], toolExecutor: ToolExecutor) {
    super();
    this.availableTools = tools;
    this.toolExecutor = toolExecutor;
  }

  // Processes the current state and decides the next action
  async think(): Promise<boolean> {
    // Validate the prompt and concatenate the user prompt
    if (this.nextStepPrompt.trim()) {
      this.messages.push(new HumanMessage(this.nextStepPrompt.trim()));
    }
    try {
      if (!this.chatModel) {
        throw new Error('Chat model not initialized');
      }
      // Call the AI model to get the tool call results
      const response = await this.chatModel.chat({
        messages: this.messages,
        systemPrompt: this.systemPrompt,
        tools: this.availableTools,
      });
      // Record the response for later use in Act
      this.toolCallResponse = response;

      const { content, toolCalls } = response;
      this.logger.log(`${this.name} thinking: ${content}`);
      this.logger.log(
        `${this.name} selected ${toolCalls?.length ?? 0} tools to use`,
      );
      // Send the thinking content to SSE
      this.emitThinking(content);

      // If no tools need to be called, the task is considered complete
      if (!toolCalls || toolCalls.length == 0) {
        // Only when no tools are called, manually record the assistant message
        this.messages.push(new AIMessage(content));
        this.state = AgentState.FINISHED;
        return false;
      }

      const toolCallInfo = toolCalls
        .map(
          (item: ToolCall) =>
            `Tool name: ${item.name}, arguments: ${JSON.stringify(item.args)}`,
        )
        .join('\n');
      this.logger.log('Tool call information:', toolCallInfo);
      // Send tool call information to SSE
      for (const item of toolCalls) {
        this.emitToolCall(item.name, JSON.stringify(item.args));
      }
      // When tools need to be called, no need to record the assistant message as it will be
      // automatically recorded during tool calls
      return true;
    } catch (err) {
      this.logger.error(`${this.name} process error:`, err);
      this.messages.push(new AIMessage(`${this.name} process error`));
      return false;
    }
  }

  async act(): Promise<string> {
    if (!this.toolCallResponse) {
      return 'No tools need to be called';
    }
    const { content, toolCalls } = this.toolCallResponse;
    if (!toolCalls || toolCalls.length === 0) {
      return 'No tools need to be called';
    }
    this.messages.push(new AIMessage({ content, tool_calls: toolCalls }));
    const results: string[] = [];

    // Call tools
    for (const toolCall of toolCalls) {
      const toolCallId = toolCall.id || crypto.randomUUID();
      try {
        const content = await this.toolExecutor.execute(
          toolCall.name,
          toolCall.args,
        );
        this.messages.push(
          new ToolMessage({
            content,
            tool_call_id: toolCallId,
            name: toolCall.name,
          }),
        );
        // Send each tool's result to SSE
        this.emitToolResult(toolCall.name, content);
        results.push(`Tool ${toolCall.name} returned content: ${content}`);
        // Check if a terminate tool was called
        if (toolCall.name === TerminateTool.name) {
          this.state = AgentState.FINISHED;
        }
      } catch (err) {
        this.logger.error(`Executing tool ${toolCall.name} error:`, err);
        const errContent = `Executing tool ${toolCall.name} error`;
        this.messages.push(
          new ToolMessage({
            content: errContent,
            tool_call_id: toolCallId,
            name: toolCall.name,
          }),
        );
        this.emitToolResult(toolCall.name, errContent);
        results.push(errContent);
      }
    }

    this.toolCallResponse = null;
    const finalResult = results.join('\n');
    this.logger.log('Final result:', finalResult);
    return finalResult;
  }
}
